mozetl/taar/taar_similarity.py

```diff
@@ -0,0 +1,285 @@
+
+        # Here a larger distance indicates a poorer match between categorical variables.
+        j_d = (distance.hamming(x_cat_f, y_cat_f))
+        j_c = (distance.canberra(x_con_f, y_con_f))
+        # Take the product of similarities to attain a univariate similarity score.
+        return abs((j_c + 0.001) * j_d)
+
+    # Generate a pairRDD for faster comparison between rows.
+    def generate_non_cartesian_pairs(rdd1, rdd2=None):
+        # If only one RDD is passed, a pair RDD is based on intra-cluster comparison.
+        if rdd2:
+            rdd2b = rdd2.zipWithIndex().map(lambda p: (p[1], p[0]))
+            rdd1b = rdd1.zipWithIndex().map(lambda p: (p[1], p[0]))
+        # If two RDDs are passed, a pair RDD is based on inter-cluster comparison.
+        else:
+            rdd_split_1, rdd_split_2 = rdd1.randomSplit([0.5, 0.5], RAND_SEED)
+            rdd2b = rdd_split_2.zipWithIndex().map(lambda p: (p[1], p[0]))
+            rdd1b = rdd_split_1.zipWithIndex().map(lambda p: (p[1], p[0]))
+        return rdd1b.join(rdd2b).map(lambda p: p[1])
+
+    # Instantiate holder lists for inter- and intra-cluster scores.
+    same_cluster_scores = []
+    different_cluster_scores = []
+
+    for cluster_number in clusters_histogram:
+        def target_clust(c_row): return c_row['prediction'] == cluster_number
+
+        def off_target_clust(c_row): return c_row['prediction'] != cluster_number
+
+        target_cluster, off_target_clusters = (features_df.filter(f) for f in (target_clust, off_target_clust))
+
+        # Compares the similarity score between pairs of clients in the same cluster.
+        pair_rdd = generate_non_cartesian_pairs(target_cluster.rdd)
+        same_clusters = pair_rdd.map(lambda p: sim_fum(p)).collect()
+        same_cluster_scores.append(same_clusters)
+
+        # Compares the similarity score between pairs of clients in different clusters.
+        pair_rdd = generate_non_cartesian_pairs(target_cluster.rdd, off_target_clusters.rdd)
+        different_clusters = pair_rdd.map(lambda p: sim_fum(p)).collect()
+        different_cluster_scores.append(different_clusters)
+
+    # Determine a range of observed similarity values linearly spaced.
+    max_sim = max(max(different_cluster_scores), max(same_cluster_scores))
+    min_sim = min(min(different_cluster_scores), min(same_cluster_scores))
+    lr_index = numpy.arange(min_sim, max_sim, float(abs(min_sim-max_sim))/1000)
+
+    # Kernel density estimate for the inter-cluster comparison scores.
+    kd_dc = KernelDensity()
+    kd_dc.setSample(spark.sparkContext.parrallelize(different_cluster_scores))
+    kd_dc.setBandwidth(KDE_KERNEL_BANDWIDTH)
+    denominator_density = kd_dc.estimate(lr_index)
+
+    # Kernel density estimate for the intra-cluster comparison scores.
+    kd_sc = KernelDensity()
+    kd_sc.setSample(spark.sparkContext.parrallelize(same_cluster_scores))
+    kd_sc.setBandwidth(KDE_KERNEL_BANDWIDTH)
+    numerator_density = kd_sc.estimate(lr_index)
```

Comment:
Test coverage should include:
1- Perform an element-wise division between the elements of `numerator_density` and `denominator_density`.
2- The number of elements should equal the number of elements in  `lr_index`
3- There should exists at least 1 value between 0.0 and 1.0
4- There should exists at least 1 value between 1.0 and infinity